packages/client/src/runtime/core/engines/client/TransactionManager.test.ts

```diff
@@ -0,0 +1,111 @@
+  TransactionTimedOutError,
+} from './TransactionManagerErrors'
+
+jest.useFakeTimers()
+
+const START_TRANSACTION_TIME = 200
+
+let transactionManager: TransactionManager
+
+beforeEach(() => {
+  const mockQueryAble: Queryable = {
+    adapterName: 'mock-adapter',
+    provider: 'postgres',
+    executeRaw: jest.fn().mockRejectedValue('Not implemented for test'),
+    queryRaw: jest.fn().mockRejectedValue('Not implemented for test'),
+  }
+
+  const mockTransaction: Transaction = {
+    ...mockQueryAble,
+    options: { usePhantomQuery: false },
+    commit: jest.fn().mockResolvedValue(ok(undefined)),
+    rollback: jest.fn().mockResolvedValue(ok(undefined)),
+  }
+
+  const transactionContext: TransactionContext = {
+    ...mockQueryAble,
+    startTransaction(): Promise<Result<Transaction>> {
+      return new Promise((resolve) =>
+        setTimeout(() => {
+          resolve(ok(mockTransaction))
+        }, START_TRANSACTION_TIME),
+      )
+    },
+  }
+
+  const driverAdapter: DriverAdapter = {
+    provider: 'postgres',
+    adapterName: '@prisma/adapter-mock',
+    queryRaw: jest.fn().mockRejectedValue('Not implemented for test'),
+    executeRaw: jest.fn().mockRejectedValue('Not implemented for test'),
+    transactionContext: () => {
+      return Promise.resolve(ok(transactionContext))
+    },
+  }
+
+  transactionManager = new TransactionManager(driverAdapter)
+})
+
+test('transaction executes normally', async () => {
+  const start = transactionManager.startTransaction({ timeout: 500, maxWait: START_TRANSACTION_TIME * 2 })
+
+  await jest.advanceTimersByTimeAsync(START_TRANSACTION_TIME + 100)
+  const { id } = await start
+
+  await transactionManager.commitTransaction(id)
+
+  await expect(transactionManager.commitTransaction(id)).rejects.toBeInstanceOf(TransactionAlreadyCommittedError)
+  await expect(transactionManager.rollbackTransaction(id)).rejects.toBeInstanceOf(TransactionAlreadyCommittedError)
+})
+
+test('transaction is rolled back', async () => {
+  const start = transactionManager.startTransaction({ timeout: 500, maxWait: START_TRANSACTION_TIME * 2 })
+
+  await jest.advanceTimersByTimeAsync(START_TRANSACTION_TIME + 100)
+  const { id } = await start
+
+  await transactionManager.rollbackTransaction(id)
+
+  await expect(transactionManager.commitTransaction(id)).rejects.toBeInstanceOf(TransactionRolledBackError)
+  await expect(transactionManager.rollbackTransaction(id)).rejects.toBeInstanceOf(TransactionRolledBackError)
+})
+
+//  TODO: this test is failing and I have no idea why...
+test('transaction times out during starting', async () => {
+  const start = transactionManager.startTransaction({ timeout: 500, maxWait: START_TRANSACTION_TIME / 2 })
+
+  await jest.advanceTimersByTimeAsync(START_TRANSACTION_TIME + 100)
+
+  // TODO: This line asserts the error but somehow the error is still thrown beyond it and the test fails???
+  await expect(start).rejects.toBeInstanceOf(TransactionTimedOutError)
+})
```

Comment:
❓ I can't really figure out why this test keeps failing... maybe someone has a good idea?